README.md
# Styled ByteStream

A library for printing rich output to a terminal using [amphp/byte-stream](https://github.com/amphp/byte-stream)!

- Supports default foreground and background colors
  - black
  - blue
  - cyan
  - green
  - magenta
  - red
  - white
  - yellow
- Supports default formatting options
  - bold
  - underline
  - reverse
  - conceal
- Other helper methods designed to make terminal output easy to create
- An immutable fluent API for creating OutputStream with a specific style

## Installation

[Composer](https://getcomposer.org/) is the only supported way to install Styled ByteStream

```
composer require cspray/labrador-styled-byte-stream
```

## Documentation

Styled ByteStream provides a single `Amp\ByteStream\OutputStream` decorator; the `Cspray\Labrador\StyledByteStream\TerminalOutputStream`. 
This object provides an API for creating rich terminal output. This includes outputting content with new lines, changing 
the color of the text, and/or adding formatting options. All of these methods can be chained together to come up with 
exactly the right style of content for your Amp powered CLI app.

All examples below assume to be running in the following boilerplate:

```php
<?php

require_once dirname(__DIR__) . '/vendor/autoload.php';

use Amp\Loop;
use Cspray\Labrador\StyledByteStream\TerminalOutputStream;
use function Amp\ByteStream\getStdout;

$stream = new TerminalOutputStream(getStdout());

// The example code is expected to be executed here
```

### Write with new line

The `writeln(string $text)` method writes some text to the decorated `OutputStream` and appends a `PHP_EOL` on the end.

```php
$stream->writeln('This is content on a new line');
```

### Add multiple line breaks

The `br(int $count = 1)` method will write `$count` number of `PHP_EOL` to the decorated `OutputStream`.

```php
$stream->br(3); // adds 3 new lines to the terminal output
```

### Background Colors

For every supported color there is a corresponding background method. For example, `backgroundBlack()`, `blackgroundBlue()`, etc. 
These methods output text with the appropriate background.

```php
$stream->backgroundBlue('This has a blue background');
$stream->backgroundYellow('This has a yellow background');
```

### Foreground Colors

Every supported color is also a method that allows setting the foreground color. For example, `black()`, `blue`, etc. 
These methods output text with the appropriate text color.

```php
$stream->magenta('This is magenta text');
$stream->cyan('This is cyan text');
```

### Formatting Options

Every supported formatting option is also a method that allows changing the way the output text is displayed. For example, 
`bold()`, `underline()`, etc. These methods output text with the appropriate formatting.

```php
$stream->bold('This is bolded text');
$stream->underline('This text has an underline');
```

### Chaining styles

All color and formatting options can be chained together to compose exactly the style you need.

```php
$stream->backgroundWhite()->red('This has a white background and red text');
$stream->bold()->yellow()->underline()->backgroundRed('The order of the chaining does not matter');
```

### Chaining styles and new lines

By default, the color and formatting options will append a new line to the end of any text passed to them. Bypass this
default behavior by chaining the `inline()` method. The `inline()` method can only be chained, text cannot be passed to 
this method.

```php
$stream->inline()->bold()->red('This is inline bold red text');
$stream->bold()->red('... This is text with a new line at the end');
```

Alternatively, skip over automatically appended lines by chaining to the `write` method.

```php
$stream->bold()->red()->write('I am bold red inline text too!');
```

It is possible to _force_ new lines... even if the `write` method is being used. Control the number of new lines by 
passing an `int` to the `forceNewline()` method. The `forceNewline` method can only be chained, text cannot be passed to 
this method.

```php
$stream->bold()->red()->forceNewline(2)->write('I am bold red text with 2 new lines at the end');
```

### Using Immutability

Chaining method calls is a straightforward approach for outputting rich text. It is important to realize that the 
`TerminalOutputStream` is an immutable object; each call to a color or formatting option will create a new instance with 
the defined formatting. This design can be taken advantage of to quickly compose reusable "styles" and have dependencies 
only be aware of the underlying `Amp\ByteStream\OutputStream` interface.

Let's assume that an implementation similar to the following is provided. Our fictional application will run a series 
of reports that can be of a state _successful_, _failed_, or _disabled_. The CLI application running the reports should 
print out a summary of which reports were processed and style the reports differently based on its state.

> The below code examples are standalone and not intended to run in the example boilerplate

```php
<?php

use Amp\ByteStream\OutputStream;
use Amp\Promise;use function Amp\call;

class ReportSummaryPrinter {

    public function __construct(private OutputStream $output) {}

    public function writeReportResults(array $report) : Promise {
        return call(function() use($report) {
            $this->output->write($report['name'] . ' received');
        });    
    }

}
```

Use `TerminalOutputStream` to create decorated `OutputStream` so the `ReportSummaryPrinter` is not complicated with 
styling logic or knowledge of the `TerminalOutputStream`. Later on if a different `OutputStream` is required, for example 
to write the summary to a file, it will only be necessary to swap out the implementations.

```php
<?php

require_once dirname(__DIR__) . '/vendor/autoload.php';

use Cspray\Labrador\StyledByteStream\TerminalOutputStream;
use function Amp\ByteStream\getStdout;

$stream = new TerminalOutputStream(getStdout());

$successfulReportOutput = $stream->forceNewline()->green();
$failedReportOutput = $stream->forceNewline()->backgroundRed()->white()->bold();
$disabledReportOutput = $stream->forceNewline()->yellow()->underline();

$successfulReportPrinter = new ReportSummaryPrinter($successfulReportOutput);
$failedReportPrinter = new ReportSummaryPrinter($failedReportOutput);
$disabledReportPrinter = new ReportSummaryPrinter($disabledReportOutput);

$successfulReportPrinter->writeReportResults(['name' => 'Foo Bar Report']);
$failedReportPrinter->writeReportResults(['name' => 'A failed report!']);
$disabledReportPrinter->writeReportResults(['name' => 'We never ran this report...']);
```

> This is a working example! If you clone the repo and run `php examples/report-summary-printer.php` you'll 
> see the expected output in your terminal!

## Governance

All Labrador packages adhere to the rules laid out in the [Labrador Governance repo](https://github.com/labrador-kennel/governance)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## 0.2.0 - 2023-02-25

### Changed

- Updated to use Amp v3 and Fibers
- Updated `TerminalOuptutStream` to implement `Amp\ByteStream\WritableStream`

## 0.1.0 - 2021-05-11

### Added

- `TerminalOutputStream` object with support for default colors and formatting options.

test/TerminalOutputStreamTest.php
<?php declare(strict_types=1);

namespace Cspray\Labrador\StyledByteStream;

use Amp\ByteStream\WritableBuffer;
use Amp\ByteStream\WritableStream;
use PHPUnit\Framework\TestCase;

class TerminalOutputStreamTest extends TestCase {

    private WritableBuffer $buffer;

    private TerminalOutputStream $subject;

    public function setUp() : void {
        $this->buffer = new WritableBuffer();
        $this->subject = new TerminalOutputStream($this->buffer);
    }

    public function normalColorMap() {
        return [
            'black' => ['black', 0],
            'red' => ['red', 1],
            'green' => ['green', 2],
            'yellow' => ['yellow', 3],
            'blue' => ['blue', 4],
            'magenta' => ['magenta', 5],
            'cyan' => ['cyan', 6],
            'white' => ['white', 7]
        ];
    }

    /**
     * @dataProvider normalColorMap
     */
    public function testBackgroundColor(string $name, int $code) : void {
        $method = 'background' . ucfirst($name);
        $this->subject->$method('text');
        $this->buffer->end();

        $expected = "\033[4" . $code . "mtext\033[49m" . PHP_EOL;
        $this->assertSame($expected, $this->buffer->buffer());
    }

    /**
     * @dataProvider normalColorMap
     */
    public function testForegroundColor(string $name, int $code) {
        $this->subject->$name('text');
        $this->buffer->end();

        $expected = "\033[3" . $code . "mtext\033[39m" . PHP_EOL;
        $this->assertSame($expected, $this->buffer->buffer());
    }

    /**
     * @dataProvider normalColorMap
     */
    public function testBoldingColors(string $name, int $code) {
        $this->subject->bold()->$name()->write('text');
        $this->buffer->end();

        $expected = "\033[1;3" . $code . "mtext\033[22;39m";
        $this->assertSame($expected, $this->buffer->buffer());
    }

    /**
     * @dataProvider normalColorMap
     */
    public function testUnderliningColors(string $name, int $code) {
        $this->subject->underline()->$name()->write('text');
        $this->buffer->end();

        $expected = "\033[4;3" . $code . "mtext\033[24;39m";
        $this->assertSame($expected, $this->buffer->buffer());
    }

    /**
     * @dataProvider normalColorMap
     */
    public function testReversingColors(string $name, int $code) {
        $this->subject->reverse()->$name()->write('text');
        $this->buffer->end();

        $expected = "\033[7;3" . $code . "mtext\033[27;39m";
        $this->assertSame($expected, $this->buffer->buffer());
    }

    /**
     * @dataProvider normalColorMap
     */
    public function testConcealingColors(string $name, int $code) {
        $this->subject->conceal()->$name()->write('text');
        $this->buffer->end();

        $expected = "\033[8;3" . $code . "mtext\033[28;39m";
        $this->assertSame($expected, $this->buffer->buffer());
    }

    /**
     * @dataProvider normalColorMap
     */
    public function testForceLineWrite(string $name, int $code) {
        $this->subject->forceNewline(5)->$name()->write('foobar');
        $this->buffer->end();

        $expected = "\033[3" . $code . "mfoobar\033[39m" . str_repeat(PHP_EOL, 5);
        $this->assertSame($expected, $this->buffer->buffer());
    }

    /**
     * @dataProvider normalColorMap
     */
    public function testForceLineWriteln(string $name, int $code) {
        $this->subject->forceNewline(5)->$name()->writeln('foobar');
        $this->buffer->end();

        $expected = "\033[3" . $code . "mfoobar\033[39m" . str_repeat(PHP_EOL, 6);
        $this->assertSame($expected, $this->buffer->buffer());
    }

    public function chainedForeGroundAndBackground() {
        return [
            // White foreground
            'whiteOnBlack' => ['white', 'backgroundBlack', 7, 0],
            'whiteOnRed' => ['white', 'backgroundRed', 7, 1],
            'whiteOnGreen' => ['white', 'backgroundGreen', 7, 2],
            'whiteOnYellow' => ['white', 'backgroundYellow', 7, 3],
            'whiteOnBlue' => ['white', 'backgroundBlue', 7, 4],
            'whiteOnMagenta' => ['white', 'backgroundMagenta', 7, 5],
            'whiteOnCyan' => ['white', 'backgroundCyan', 7, 6],
            'whiteOnWhite' => ['white', 'backgroundWhite', 7, 7],

            // Cyan foreground
            'cyanOnBlack' => ['cyan', 'backgroundBlack', 6, 0],
            'cyanOnRed' => ['cyan', 'backgroundRed', 6, 1],
            'cyanOnGreen' => ['cyan', 'backgroundGreen', 6, 2],
            'cyanOnYellow' => ['cyan', 'backgroundYellow', 6, 3],
            'cyanOnBlue' => ['cyan', 'backgroundBlue', 6, 4],
            'cyanOnMagenta' => ['cyan', 'backgroundMagenta', 6, 5],
            'cyanOnCyan' => ['cyan', 'backgroundCyan', 6, 6],
            'cyanOnWhite' => ['cyan', 'backgroundWhite', 6, 7],

            // Magenta foreground
            'magentaOnBlack' => ['magenta', 'backgroundBlack', 5, 0],
            'magentaOnRed' => ['magenta', 'backgroundRed', 5, 1],
            'magentaOnGreen' => ['magenta', 'backgroundGreen', 5, 2],
            'magentaOnYellow' => ['magenta', 'backgroundYellow', 5, 3],
            'magentaOnBlue' => ['magenta', 'backgroundBlue', 5, 4],
            'magentaOnMagenta' => ['magenta', 'backgroundMagenta', 5, 5],
            'magentaOnCyan' => ['magenta', 'backgroundCyan', 5, 6],
            'magentaOnWhite' => ['magenta', 'backgroundWhite', 5, 7],

            // Blue foreground
            'blueOnBlack' => ['blue', 'backgroundBlack', 4, 0],
            'blueOnRed' => ['blue', 'backgroundRed', 4, 1],
            'blueOnGreen' => ['blue', 'backgroundGreen', 4, 2],
            'blueOnYellow' => ['blue', 'backgroundYellow', 4, 3],
            'blueOnBlue' => ['blue', 'backgroundBlue', 4, 4],
            'blueOnMagenta' => ['blue', 'backgroundMagenta', 4, 5],
            'blueOnCyan' => ['blue', 'backgroundCyan', 4, 6],
            'blueOnWhite' => ['blue', 'backgroundWhite', 4, 7],

            // Yellow foreground
            'yellowOnBlack' => ['yellow', 'backgroundBlack', 3, 0],
            'yellowOnRed' => ['yellow', 'backgroundRed', 3, 1],
            'yellowOnGreen' => ['yellow', 'backgroundGreen', 3, 2],
            'yellowOnYellow' => ['yellow', 'backgroundYellow', 3, 3],
            'yellowOnBlue' => ['yellow', 'backgroundBlue', 3, 4],
            'yellowOnMagenta' => ['yellow', 'backgroundMagenta', 3, 5],
            'yellowOnCyan' => ['yellow', 'backgroundCyan', 3, 6],
            'yellowOnWhite' => ['yellow', 'backgroundWhite', 3, 7],

            // Green foreground
            'greenOnBlack' => ['green', 'backgroundBlack', 2, 0],
            'greenOnRed' => ['green', 'backgroundRed', 2, 1],
            'greenOnGreen' => ['green', 'backgroundGreen', 2, 2],
            'greenOnYellow' => ['green', 'backgroundYellow', 2, 3],
            'greenOnBlue' => ['green', 'backgroundBlue', 2, 4],
            'greenOnMagenta' => ['green', 'backgroundMagenta', 2, 5],
            'greenOnCyan' => ['green', 'backgroundCyan', 2, 6],
            'greenOnWhite' => ['green', 'backgroundWhite', 2, 7],

            // Red foreground
            'redOnBlack' => ['red', 'backgroundBlack', 1, 0],
            'redOnRed' => ['red', 'backgroundRed', 1, 1],
            'redOnGreen' => ['red', 'backgroundGreen', 1, 2],
            'redOnYellow' => ['red', 'backgroundYellow', 1, 3],
            'redOnBlue' => ['red', 'backgroundBlue', 1, 4],
            'redOnMagenta' => ['red', 'backgroundMagenta', 1, 5],
            'redOnCyan' => ['red', 'backgroundCyan', 1, 6],
            'redOnWhite' => ['red', 'backgroundWhite', 1, 7],

            // Black foreground
            'blackOnBlack' => ['black', 'backgroundBlack', 0, 0],
            'blackOnRed' => ['black', 'backgroundRed', 0, 1],
            'blackOnGreen' => ['black', 'backgroundGreen', 0, 2],
            'blackOnYellow' => ['black', 'backgroundYellow', 0, 3],
            'blackOnBlue' => ['black', 'backgroundBlue', 0, 4],
            'blackOnMagenta' => ['black', 'backgroundMagenta', 0, 5],
            'blackOnCyan' => ['black', 'backgroundCyan', 0, 6],
            'blackOnWhite' => ['black', 'backgroundWhite', 0, 7],
        ];
    }

    /**
     * @dataProvider chainedForeGroundAndBackground
     */
    public function testChainingForegroundAndBackgroundWriteln(string $foreground, string $background, int $foregroundCode, int $backgroundCode) {
        $this->subject->$background()->$foreground()->writeln('text');
        $this->buffer->end();
        $expected = "\033[4" . $backgroundCode . ";3" . $foregroundCode . "mtext\033[49;39m" . PHP_EOL;
        $this->assertSame($expected, $this->buffer->buffer());
    }

    /**
     * @dataProvider chainedForeGroundAndBackground
     */
    public function testChainingForegroundAndBackgroundWrite(string $foreground, string $background, int $foregroundCode, int $backgroundCode) {
        $this->subject->$background()->$foreground()->write('text');
        $this->buffer->end();
        $expected = "\033[4" . $backgroundCode . ";3" . $foregroundCode . "mtext\033[49;39m";
        $this->assertSame($expected, $this->buffer->buffer());
    }

    /**
     * @dataProvider chainedForeGroundAndBackground
     */
    public function testChainingForeGroundAndBackgroundEndsOnForeGround(string $foreground, string $background, int $foregroundCode, int $backgroundCode) {
        $this->subject->$background()->$foreground('text');
        $this->buffer->end();
        $expected = "\033[4" . $backgroundCode . ";3" . $foregroundCode . "mtext\033[49;39m" . PHP_EOL;
        $this->assertSame($expected, $this->buffer->buffer());
    }

    /**
     * @dataProvider chainedForeGroundAndBackground
     */
    public function testChainingForeGroundAndBackgroundEndsOnBackGround(string $foreground, string $background, int $foregroundCode, int $backgroundCode) {
        $this->subject->$foreground()->$background('text');
        $this->buffer->end();
        $expected = "\033[3" . $foregroundCode . ";4" . $backgroundCode . "mtext\033[39;49m" . PHP_EOL;
        $this->assertSame($expected, $this->buffer->buffer());
    }

    /**
     * @dataProvider chainedForeGroundAndBackground
     */
    public function testChainedForegroundAndBackgroundInlineEndOnBackground(string $foreground, string $background, int $foregroundCode, int $backgroundCode) {
        $this->subject->$foreground()->inline()->$background('text');
        $this->buffer->end();
        $expected = "\033[3" . $foregroundCode . ";4" . $backgroundCode . "mtext\033[39;49m";
        $this->assertSame($expected, $this->buffer->buffer());
    }

    /**
     * @dataProvider chainedForeGroundAndBackground
     */
    public function testChainedForegroundAndBackgroundInlineEndOnForeground(string $foreground, string $background, int $foregroundCode, int $backgroundCode) {
        $this->subject->$background()->inline()->$foreground('text');
        $this->buffer->end();
        $expected = "\033[4" . $backgroundCode . ";3" . $foregroundCode . "mtext\033[49;39m";
        $this->assertSame($expected, $this->buffer->buffer());
    }

    public function formattingOptionsMap() {
        return [
            'bold' => ['bold', 1, 22],
            'underline' => ['underline', 4, 24],
            'reverse' => ['reverse', 7, 27],
            'conceal' => ['conceal', 8, 28]
        ];
    }

    /**
     * @dataProvider formattingOptionsMap
     */
    public function testFormatting(string $method, int $set, int $unset) {
        $this->subject->$method('foo');
        $this->buffer->end();

        $expected = sprintf(
            "\033[%smfoo\033[%sm" . PHP_EOL,
            $set,
            $unset
        );
        $this->assertSame($expected, $this->buffer->buffer());
    }

    /**
     * @dataProvider formattingOptionsMap
     */
    public function testInlineFormatting(string $method, int $set, int $unset) {
        $this->subject->inline()->$method('foo');
        $this->buffer->end();

        $expected = sprintf(
            "\033[%smfoo\033[%sm",
            $set,
            $unset
        );
        $this->assertSame($expected, $this->buffer->buffer());
    }

    public function testChainedAllFormatting() {
        $this->subject->bold()->underline()->reverse()->conceal('foobar');
        $this->buffer->end();
        $expected = "\033[1;4;7;8mfoobar\033[22;24;27;28m" . PHP_EOL;
        $this->assertSame($expected, $this->buffer->buffer());
    }

    public function testWriteLn() {
        $this->subject->writeln('AsyncUnit');
        $this->buffer->end();

        $expected = 'AsyncUnit' . PHP_EOL;
        $this->assertSame($expected, $this->buffer->buffer());
    }

    public function testBrNoArgs() {
        $this->subject->br();
        $this->buffer->end();

        $this->assertSame(PHP_EOL, $this->buffer->buffer());
    }

    public function testBrArgs() {
        $randomInt = random_int(1, 20);
        $this->subject->br($randomInt);
        $this->buffer->end();

        $this->assertSame(str_repeat(PHP_EOL, $randomInt), $this->buffer->buffer());
    }

    public function testEnding() {
        $this->subject->end('the end');

        $this->assertSame('the end', $this->buffer->buffer());
    }

    public function testCloseMethodDelegatedToDecoratedStream() : void {
        $mock = $this->getMockBuilder(WritableStream::class)->getMock();
        $subject = new TerminalOutputStream($mock);

        $mock->expects($this->once())->method('close');

        $subject->close();
    }

    public function testIsClosedMethodDelegatedToDecoratedStream() : void {
        $mock = $this->getMockBuilder(WritableStream::class)->getMock();
        $subject = new TerminalOutputStream($mock);

        $mock->expects($this->once())->method('isClosed')->willReturn(true);

        self::assertTrue($subject->isClosed());
    }

    public function testIsWritableMethodDelegatedToDecoratedStream() : void {
        $mock = $this->getMockBuilder(WritableStream::class)->getMock();
        $subject = new TerminalOutputStream($mock);

        $mock->expects($this->once())->method('isWritable')->willReturn(true);

        self::assertTrue($subject->isWritable());
    }

    public function testOnCloseMethodDelegatedToDecoratedStream() : void {
        $mock = $this->getMockBuilder(WritableStream::class)->getMock();
        $subject = new TerminalOutputStream($mock);

        $closure = static function() {};
        $mock->expects($this->once())->method('onClose')->with($closure);

        $subject->onClose($closure);
    }

}

examples/report-summary-printer.php
<?php declare(strict_types=1);

require_once dirname(__DIR__) . '/vendor/autoload.php';

use Amp\ByteStream\WritableStream;
use Cspray\Labrador\StyledByteStream\TerminalOutputStream;
use function Amp\ByteStream\getStdout;

function createSummaryPrinter(WritableStream $output) : object {
    return new class($output) {

        public function __construct(private WritableStream $output) {}

        public function writeReportResults(array $report) : void {
            $this->output->write($report['name'] . ' received');
        }

    };
}

$stream = new TerminalOutputStream(getStdout());

$successfulReportOutput = $stream->forceNewline()->green();
$failedReportOutput = $stream->forceNewline()->backgroundRed()->white()->bold();
$disabledReportOutput = $stream->forceNewline()->yellow()->underline();

$successfulReportPrinter = createSummaryPrinter($successfulReportOutput);
$failedReportPrinter = createSummaryPrinter($failedReportOutput);
$disabledReportPrinter = createSummaryPrinter($disabledReportOutput);

$successfulReportPrinter->writeReportResults(['name' => 'Foo Bar Report']);
$failedReportPrinter->writeReportResults(['name' => 'Bad Data Report']);
$disabledReportPrinter->writeReportResults(['name' => 'Old Legacy Report']);


src/TerminalOutputStream.php
<?php declare(strict_types=1);

namespace Cspray\Labrador\StyledByteStream;

use Amp\ByteStream\WritableStream;

final class TerminalOutputStream implements WritableStream {

    private const FOREGROUND_UNSET_CODE = '39';
    private const BACKGROUND_UNSET_CODE = '49';

    private const BACKGROUND_BLACK_SET_CODE = '40';
    private const BLACK_SET_CODE = '30';

    private const BACKGROUND_RED_SET_CODE = '41';
    private const RED_SET_CODE = '31';

    private const BACKGROUND_GREEN_SET_CODE = '42';
    private const GREEN_SET_CODE = '32';

    private const BACKGROUND_YELLOW_SET_CODE = '43';
    private const YELLOW_SET_CODE = '33';

    private const BACKGROUND_BLUE_SET_CODE = '44';
    private const BLUE_SET_CODE = '34';

    private const BACKGROUND_MAGENTA_SET_CODE = '45';
    private const MAGENTA_SET_CODE = '35';

    private const BACKGROUND_CYAN_SET_CODE = '46';
    private const CYAN_SET_CODE = '36';

    private const BACKGROUND_WHITE_SET_CODE = '47';
    private const WHITE_SET_CODE = '37';

    private const BOLD_SET_CODE = '1';
    private const BOLD_UNSET_CODE = '22';

    private const UNDERLINE_SET_CODE = '4';
    private const UNDERLINE_UNSET_CODE = '24';

    private const REVERSE_SET_CODE = '7';
    private const REVERSE_UNSET_CODE = '27';

    private const CONCEAL_SET_CODE = '8';
    private const CONCEAL_UNSET_CODE = '28';

    private array $set = [];
    private array $unset = [];

    private bool $inline = false;
    private int $forceLineCount = 0;

    public function __construct(private WritableStream $output) {}

    public function write(string $data) : void {
        $this->doWrite($data);
    }

    public function end(string $finalData = null) : void {
        if ($finalData !== null) {
            $this->output->write($finalData);
        }
        $this->output->end();
    }

    public function writeln(string $data) : void {
        $this->doWriteLn($data);
    }

    public function br(int $count = 1) : void {
        $this->write(str_repeat(PHP_EOL, $count));
    }

    public function inline() : TerminalOutputStream {
        $newStream = clone $this;
        $newStream->inline = true;
        return $newStream;
    }

    public function forceNewline(int $count = 1) : TerminalOutputStream {
        $newStream = clone $this;
        $newStream->forceLineCount = $count;
        return $newStream;
    }

    public function backgroundBlack(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::BACKGROUND_BLACK_SET_CODE, self::BACKGROUND_UNSET_CODE);
    }

    public function black(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::BLACK_SET_CODE, self::FOREGROUND_UNSET_CODE);
    }

    public function backgroundRed(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::BACKGROUND_RED_SET_CODE, self::BACKGROUND_UNSET_CODE);
    }

    public function red(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::RED_SET_CODE, self::FOREGROUND_UNSET_CODE);
    }

    public function backgroundGreen(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::BACKGROUND_GREEN_SET_CODE, self::BACKGROUND_UNSET_CODE);
    }

    public function green(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::GREEN_SET_CODE, self::FOREGROUND_UNSET_CODE);
    }

    public function backgroundYellow(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::BACKGROUND_YELLOW_SET_CODE, self::BACKGROUND_UNSET_CODE);
    }

    public function yellow(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::YELLOW_SET_CODE, self::FOREGROUND_UNSET_CODE);
    }

    public function backgroundBlue(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::BACKGROUND_BLUE_SET_CODE, self::BACKGROUND_UNSET_CODE);
    }

    public function blue(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::BLUE_SET_CODE, self::FOREGROUND_UNSET_CODE);
    }

    public function backgroundMagenta(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::BACKGROUND_MAGENTA_SET_CODE, self::BACKGROUND_UNSET_CODE);
    }

    public function magenta(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::MAGENTA_SET_CODE, self::FOREGROUND_UNSET_CODE);
    }

    public function backgroundCyan(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::BACKGROUND_CYAN_SET_CODE, self::BACKGROUND_UNSET_CODE);
    }

    public function cyan(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::CYAN_SET_CODE, self::FOREGROUND_UNSET_CODE);
    }

    public function backgroundWhite(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::BACKGROUND_WHITE_SET_CODE, self::BACKGROUND_UNSET_CODE);
    }

    public function white(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::WHITE_SET_CODE, self::FOREGROUND_UNSET_CODE);
    }

    public function bold(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::BOLD_SET_CODE, self::BOLD_UNSET_CODE);
    }

    public function underline(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::UNDERLINE_SET_CODE, self::UNDERLINE_UNSET_CODE);
    }

    public function reverse(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::REVERSE_SET_CODE, self::REVERSE_UNSET_CODE);
    }

    public function conceal(string $text = null) : TerminalOutputStream {
        return $this->handle($text, self::CONCEAL_SET_CODE, self::CONCEAL_UNSET_CODE);
    }

    private function handle(?string $text, string $setCode, string $unsetCode) : TerminalOutputStream {
        if (is_null($text)) {
            $newStream = clone $this;
            $newStream->set[] = $setCode;
            $newStream->unset[] = $unsetCode;
            return $newStream;
        }

        $method = $this->inline ? 'doWrite' : 'doWriteln';
        return $this->$method($text, [$setCode], [$unsetCode]);
    }

    private function doWriteLn(string $text, array $set = [], array $unset = []) : TerminalOutputStream {
        $this->output->write($this->getFormattedText($text, $set, $unset, PHP_EOL));
        return $this;
    }

    private function doWrite(string $text, array $set = [], array $unset = []) : TerminalOutputStream {
        $this->output->write($this->getFormattedText($text, $set, $unset));
        return $this;
    }

    private function getFormattedText(string $text, array $set, array $unset, string $suffix = '') : string {
        $finalSet = array_merge([], $this->set, $set);
        $finalUnset = array_merge([], $this->unset, $unset);
        if ($this->forceLineCount > 0) {
            $suffix .= str_repeat(PHP_EOL, $this->forceLineCount);
        }
        if (empty($finalSet)) {
            return $text . $suffix;
        }

        return sprintf(
            "\033[%sm%s\033[%sm%s",
            implode(';', $finalSet),
            $text,
            implode(';', $finalUnset),
            $suffix
        );
    }

    public function close() : void {
        $this->output->close();
    }

    public function isClosed() : bool {
        return $this->output->isClosed();
    }

    public function onClose(\Closure $onClose) : void {
        $this->output->onClose($onClose);
    }

    public function isWritable() : bool {
        return $this->output->isWritable();
    }
}